server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreShardAllocationIT.java

```diff
@@ -0,0 +1,463 @@
+    }
+
+    private void initializeCluster () {
+        internalCluster().setBootstrapClusterManagerNodeIndex(0);
+        List<String> cmNodes = internalCluster().startNodes(1);
+        cmNodeName = cmNodes.get(0);
+        client = internalCluster().client();
+    }
+
+    private void setClusterMode (String mode) {
+        updateSettingsRequest.persistentSettings(Settings.builder().put(REMOTE_STORE_COMPATIBILITY_MODE_SETTING.getKey(), mode));
+        assertAcked(client.admin().cluster().updateSettings(updateSettingsRequest).actionGet());
+    }
+
+    private void setDirection (String direction) {
+        updateSettingsRequest.persistentSettings(Settings.builder().put(DIRECTION_SETTING.getKey(), direction));
+        assertAcked(client.admin().cluster().updateSettings(updateSettingsRequest).actionGet());
+    }
+
+    private DiscoveryNode assertNodeInCluster (String nodeName) {
+        Map<String, DiscoveryNode> nodes = client.admin().cluster().prepareState().get().getState().nodes().getNodes();
+        DiscoveryNode discoveryNode = null;
+        for (Map.Entry<String, DiscoveryNode> entry : nodes.entrySet()) {
+            DiscoveryNode node = entry.getValue();
+            if (node.getName().equals(nodeName)) {
+                discoveryNode = node;
+                break;
+            }
+        }
+        assertNotNull(discoveryNode);
+        return discoveryNode;
+    }
+
+    public void testDontAllocateNewPrimaryShardOnNonRemoteNodeForRemoteStoreDirection () throws Exception {
+        logger.info(" --> initialize cluster");
+        initializeCluster();
+
+        logger.info(" --> add non-remote node");
+        addRemote = false;
+        String nonRemoteNodeName = internalCluster().startNode();
+        internalCluster().validateClusterFormed();
+        DiscoveryNode nonRemoteNode = assertNodeInCluster(nonRemoteNodeName);
+        assertFalse(nonRemoteNode.isRemoteStoreNode());
+
+        logger.info(" --> set remote_store direction");
+        setDirection("remote_store");
+
+        logger.info(" --> allocate primary shard");
+        client.admin()
+            .indices()
+            .prepareCreate(TEST_INDEX)
+            .setSettings(
+                Settings.builder()
+                    .put("index.number_of_shards", 1)
+                    .put("index.number_of_replicas", 0)
+                    .put("index.routing.allocation.include._name", nonRemoteNodeName)
+                    .put("index.routing.allocation.exclude._name", cmNodeName)
+            )
+            .execute()
+            .actionGet();
+
+        logger.info(" --> verify non-allocation of primary shard");
+        RoutingTable routingTable = client.admin().cluster().prepareState().execute().actionGet().getState().getRoutingTable();
+        ShardRouting primaryShardRouting = routingTable.index(TEST_INDEX).shard(0).primaryShard();
+        assertFalse(primaryShardRouting.active());
```

Comment:
Can we also add some extra set of assertions for this test:

- Cluster health should be RED (since primary shard is not allocated yet)
- Run the `_cluster/allocation/explain` API and assert on the reason for shard allocation. The reason should match the String we publish from the decider on a `NO` decision